pkg/objstore/s3/s3.go

```diff
@@ -161,36 +161,57 @@ func NewBucket(logger log.Logger, conf []byte, component string) (*Bucket, error
 	return NewBucketWithConfig(logger, config, component)
 }
 
+type overrideSignerType struct {
+	signerType credentials.SignatureType
+	upstream   credentials.Provider
+}
+
+func (s *overrideSignerType) Retrieve() (credentials.Value, error) {
+	v, err := s.upstream.Retrieve()
+	if err != nil {
+		return v, err
+	}
+	if !v.SignerType.IsAnonymous() {
+		v.SignerType = s.signerType
+	}
+	return v, nil
+}
+
+func (s *overrideSignerType) IsExpired() bool {
+	return s.upstream.IsExpired()
+}
+
 // NewBucketWithConfig returns a new Bucket using the provided s3 config values.
 func NewBucketWithConfig(logger log.Logger, config Config, component string) (*Bucket, error) {
 	var chain []credentials.Provider
 
+	wrapCredentialsProvider := func(p credentials.Provider) credentials.Provider {
+		if config.SignatureV2 {
+			return &overrideSignerType{signerType: credentials.SignatureV2, upstream: p}
+		}
+		return p
+	}
+
```

Comment:
What about something less dynamic? (: 
```suggestion
        wrapCredentialProvider := func(p credentials.Provider) credentials.Provider { return p }
        if config.SignatureV2 {
	wrapCredentialsProvider = func(p credentials.Provider) credentials.Provider { return &overrideSignerType{signerType: credentials.SignatureV2, upstream: p}
	}
}
```

Reply:
Is in the latest push :+1: